### PR TITLE
bpo-34574: Prevent OrderedDict iterators from exhaustion during pickling.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://bugs.python.org/issue34574

### CODE REVIEW

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_ordered_dict.py [hunk 1]
```diff
@@ -732,6 +732,26 @@ def test_key_change_during_iteration(self):
...
+        OrderedDict = self.OrderedDict
+        pairs = [('c', 1), ('b', 2), ('a', 3), ('d', 4), ('e', 5), ('f', 6)]
+        od = OrderedDict(pairs)
+
+        expected = (
+            ('keys', [k for k, v in pairs[:1]]),
+            ('values', [v for k, v in pairs[:1]]),
+            ('items', pairs[:1]),
+        )
+        for method_name, items in expected:
```
Comment (serhiy-storchaka) [inline Lib/test/test_ordered_dict.py]: Seems `items` is not used.

File: Lib/test/test_ordered_dict.py [hunk 2]
```diff
@@ -732,6 +732,26 @@ def test_key_change_during_iteration(self):
...
+            meth = getattr(od, method_name)
+            with self.subTest(method_name=method_name):
+                for i in range(pickle.HIGHEST_PROTOCOL + 1):
+                    it = iter(meth())
+                    next(it)
+                    p = pickle.dumps(it, i)
+                    unpickled = pickle.loads(p)
+                    self.assertEqual(list(it), list(unpickled))
```
Comment (serhiy-storchaka) [inline Lib/test/test_ordered_dict.py]: This test will be passed if the original iterator is exhausted and an empty list is pickled.

I would test:
```python
expected = list(meth())[1:]
self.assertEqual(list(it), expected)
self.assertEqual(list(it), expected)
```

File: Objects/odictobject.c [hunk 3]
```diff
@@ -1837,38 +1837,17 @@ PyDoc_STRVAR(reduce_doc, "Return state information for pickling");
...
 odictiter_reduce(odictiterobject *di)
 {
-    PyObject *list, *iter;
-
-    list = PyList_New(0);
-    if (!list)
-        return NULL;
+    /* copy the iterator state */
+    odictiterobject tmp = *di;
+    Py_XINCREF(tmp.di_odict);
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c:1842]: This is not enough. The state includes a reference to the result tuple.

Comment (sir-sigurd) [reply Objects/odictobject.c:1842]: It looks like the problem wasn't caused by the result tuple.

**Comment by vstinner** [comment]
@serhiy-storchaka: you approved this change, but then asked a question. It's unclear to me if you fully approved this change...

@pablogsal: Would you mind to review it?

**Pull request merged**